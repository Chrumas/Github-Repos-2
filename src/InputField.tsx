import React from "react";
import { useState, useEffect, useRef } from "react";
import { StyledInput } from "./styles/InputField.styled";
import { SearchSuggestions } from "./SearchSuggestions";
import { useSearchUpdateContext } from "./ContextProvider";

//Component responsible for processing the input data from user.
export const InputField = () => {
  const [searchTerm, setSearchTerm] = useState<string>("");
  const setSearch = useSearchUpdateContext();
  const textInput = useRef<HTMLInputElement>(null);
  const suggestionsRef = useRef<any>(null);
  const [isOpen, setIsOpen] = useState(false);
  let styleHelper = "hidden";

  //Makes sure that the dropdown menu closes when a click outside happens
  useEffect(() => {
    const closeDropdown = (e: any) => {
      var suggestionClicked: boolean = false;
      suggestionsRef.current.refCallback.current.forEach((element: any) => {
        if (e.path[0] === element) suggestionClicked = true;
      });
      if (e.path[0] === textInput.current) {
        setIsOpen(true);
      }
      if (e.path[0] !== textInput.current && !suggestionClicked) {
        if (textInput.current !== null) textInput.current.blur();
        setIsOpen(false);
      }
    };
    document.body.addEventListener("mousedown", closeDropdown);
    return () => document.body.removeEventListener("mousedown", closeDropdown);
  }, []);

  const FocusOnSearchComponent = () => {
    setSearch(searchTerm);
    setIsOpen(true);
    if (textInput.current !== null) {
      textInput.current.focus();
    }
  };

  if (document.activeElement === textInput.current && isOpen !== true) {
    setIsOpen(true);
  }

  if (isOpen) styleHelper = "visible";

  const style = { visibility: styleHelper } as React.CSSProperties;

  useEffect(() => {
    //setSearch(searchTerm);
  }, [searchTerm]);

  return (
    <StyledInput>
      <div>
        <div className="input">
          <div className="input-form">
            <input
              data-testid="input"
              ref={textInput}
              type="text"
              //value={copy}
              onChange={(e) => {
                setSearchTerm(e.target.value);
              }}
              required
            />
            <span data-testid="placeholder" className="placeholder">
              Username
            </span>
          </div>
          <div
            data-testid="button"
            className="search-button"
            onClick={() => FocusOnSearchComponent()}
          >
            Search
          </div>
        </div>
      </div>
      <div style={style} className="suggestions">
        <SearchSuggestions ref={suggestionsRef} />
      </div>
    </StyledInput>
  );
};

export default InputField;
