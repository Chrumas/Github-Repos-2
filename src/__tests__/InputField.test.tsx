import * as React from "react";
import * as ReactDOM from "react-dom";
import { InputField } from "../InputField";
import { render } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import renderer from "react-test-renderer";

it("renders without crashing", () => {
  const div = document.createElement("div");
  ReactDOM.render(<InputField />, div);
});

it("renders placeholder correctly", () => {
  const { getByTestId } = render(<InputField />);
  // eslint-disable-next-line testing-library/prefer-screen-queries
  expect(getByTestId("placeholder")).toHaveTextContent("Username");
});

it("renders button correctly", () => {
  const { getByTestId } = render(<InputField />);
  // eslint-disable-next-line testing-library/prefer-screen-queries
  expect(getByTestId("button")).toHaveTextContent("Search");
});

it("renders button with correct color", () => {
  const { getByTestId } = render(<InputField />);
  // eslint-disable-next-line testing-library/prefer-screen-queries
  expect(getByTestId("button")).toHaveStyle("background-color: #0a1236ea");
});

it("matches snapshot", () => {
  const tree = renderer.create(<InputField />).toJSON();
  expect(tree).toMatchSnapshot();
});
