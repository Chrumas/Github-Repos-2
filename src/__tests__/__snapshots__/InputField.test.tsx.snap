// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="sc-bdvvtL jgrgcp"
>
  <div>
    <div
      className="input"
    >
      <div
        className="input-form"
      >
        <input
          data-testid="input"
          onChange={[Function]}
          required={true}
          type="text"
        />
        <span
          className="placeholder"
          data-testid="placeholder"
        >
          Username
        </span>
      </div>
      <div
        className="search-button"
        data-testid="button"
        onClick={[Function]}
      >
        Search
      </div>
    </div>
  </div>
  <div
    className="suggestions"
    style={
      Object {
        "visibility": "hidden",
      }
    }
  >
    <div
      className="sc-gsDKAQ fNpFzQ"
    >
      <div
        className="search-suggestions-map-wrapper"
      >
        <div
          className="search-suggestions-map"
        >
          
        </div>
      </div>
    </div>
  </div>
</div>
`;
